### PR TITLE
feat(nuget): allow detecting source URLs via package contents

### PR DESCRIPTION
**Pull request by fgreinacher**
## Changes

Adapts the NuGet datasource to determine package source URLs from the `nupkg` file in cases where the API does not provide the [PackageBaseAddress resource](https://learn.microsoft.com/en-us/nuget/api/package-base-address-resource)

## Context

https://github.com/renovatebot/renovate/discussions/28041

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

:hammer_and_pick: with :heart: by [Siemens](https://opensource.siemens.com/)

### CODE REVIEW

File: lib/modules/datasource/nuget/types.ts [hunk 1]
```diff
@@ -10,6 +10,7 @@ export interface CatalogEntry {
   published?: string;
   projectUrl?: string;
   listed?: boolean;
+  packageContent?: string;
```
Comment (fgreinacher) [inline lib/modules/datasource/nuget/types.ts:13]: See https://learn.microsoft.com/en-us/nuget/api/registration-base-url-resource#catalog-entry

File: lib/modules/datasource/nuget/v3.ts [hunk 2]
```diff
@@ -242,3 +267,31 @@ export async function getReleases(
...
+      `${packageName}.${packageVersion}`,
+    );
+    await extract(nupkgFile, { dir: contentsDir });
+    const nuspecFile = upath.join(contentsDir, `${packageName}.nuspec`);
+    const nuspec = new XmlDocument(await fs.readCacheFile(nuspecFile, 'utf8'));
+    return nuspec.valueWithPath('metadata.repository@url');
+  } finally {
+    await fs.rmCache(cacheDir);
+  }
+}
```
Comment (fgreinacher) [inline lib/modules/datasource/nuget/v3.ts:297]: I see a couple of options for caching:

1. cache at the level of HTTP call the mapping for nupkgUrl to nupkgFile
2. cache at the level of this function the mapping from nupkgUrl to sourceUrl
3. cache one level above the mapping from (packageName, packageVersion) to sourceUrl

I'd slightly lean towards option 3, because it also optimizes the existing logic. WDYT?

**Comment by fgreinacher** [comment]
Do you have a preference whether this should be a proper configuration option or an experimental env var?

**Comment by fgreinacher** [comment]
@rarkins @viceice I have some questions, would be great if you could quickly share your opinions before I continue here 🙇

File: lib/modules/datasource/nuget/__fixtures__/nlog/NLog.4.7.3.nupkg
Comment (viceice) [inline lib/modules/datasource/nuget/__fixtures__/nlog/NLog.4.7.3.nupkg:1]: does it only contain the nuspec file? if not you should remove all unrelated files to reduce size

File: lib/modules/datasource/nuget/v3.ts [hunk 3]
```diff
@@ -242,3 +267,31 @@ export async function getReleases(
...
   return dep;
 }
+
+async function getSourceUrlFromNupkg(
+  http: Http,
+  packageName: string,
+  packageVersion: string | null,
+  nupkgUrl: string,
+): Promise<string | undefined> {
+  const cacheDir = await ensureCacheDir(`nuget`);
```
Comment (viceice) [inline lib/modules/datasource/nuget/v3.ts:277]: add an experimental environment variable to disable this. as told before I need this disabled.

File: lib/modules/datasource/nuget/v3.ts [hunk 4]
```diff
@@ -242,3 +267,31 @@ export async function getReleases(
 
   return dep;
 }
+
+async function getSourceUrlFromNupkg(
```
Comment (viceice) [inline lib/modules/datasource/nuget/v3.ts:271]: add the cache decorator to this function with a high cache time

File: lib/modules/datasource/nuget/v3.ts (hunk 2 above)
Comment (viceice) [reply lib/modules/datasource/nuget/v3.ts:297]: see above

File: lib/modules/datasource/nuget/__fixtures__/nlog/NLog.4.7.3.nupkg
Comment (fgreinacher) [reply lib/modules/datasource/nuget/__fixtures__/nlog/NLog.4.7.3.nupkg:1]: Yes, just the nuspec!

File: lib/modules/datasource/nuget/v3.ts (hunk 4 above)
Comment (fgreinacher) [reply lib/modules/datasource/nuget/v3.ts:271]: This is a function 😬

**Comment by fgreinacher** [comment]
OK, this is now ready for review. I have verified it with our internal Artifactory and also also using my public test Artifactory: https://github.com/fgreinacher/renovate-artifactory-source-url-fixed/pulls

File: lib/modules/datasource/nuget/v3.ts (hunk 4 above)
Comment (viceice) [reply lib/modules/datasource/nuget/v3.ts:297]: why not using our cache decorator?

Comment (fgreinacher) [reply lib/modules/datasource/nuget/v3.ts:297]: You cannot use decorators on functions 😬

Comment (viceice) [reply lib/modules/datasource/nuget/v3.ts:297]: declare it as class member of the datasource, there you can use the decorator

File: docs/usage/self-hosted-experimental.md [hunk 5]
```diff
@@ -141,6 +141,10 @@ This feature is in private beta.
 If set, Renovate will query the merge-confidence JSON API only for datasources that are part of this list.
 The expected value for this environment variable is a JSON array of strings.
 
+## `RENOVATE_X_NUGET_DISABLE_NUPKG_DOWNLOAD`
+
+If set, Renovate will not download `nupkg` files for determining package metadata.
+
```
Comment (rarkins) [inline docs/usage/self-hosted-experimental.md:147]: Didn't we agree to make this opt-in instead of opt-out?

File: lib/modules/datasource/nuget/v3.ts [hunk 6]
```diff
@@ -212,6 +223,20 @@ export async function getReleases(
...
+        pkgName,
+        latestStable,
+        latestNupkgUrl,
+      );
+      if (sourceUrl) {
+        dep.sourceUrl = massageUrl(sourceUrl);
+        logger.debug(
+          { sourceUrl, nupkgUrl: latestNupkgUrl },
+          `Determined sourceUrl from nupkgUrl`,
+        );
```
Comment (rarkins) [inline lib/modules/datasource/nuget/v3.ts:238]:
```suggestion
        logger.debug(
          `Determined sourceUrl ${sourceUrl} from ${latestNupkgUrl}`,
        );
```

File: lib/modules/datasource/nuget/v3.ts [hunk 7]
```diff
@@ -242,3 +267,65 @@ export async function getReleases(
...
+  packageVersion: string | null,
+  nupkgUrl: string,
+): Promise<string | null> {
+  // istanbul ignore if: experimental feature
+  if (process.env.RENOVATE_X_NUGET_DISABLE_NUPKG_DOWNLOAD) {
+    logger.debug(
+      `Skipping nupkg download because RENOVATE_X_NUGET_DISABLE_NUPKG_DOWNLOAD is set.`,
+    );
+    return null;
+  }
```
Comment (rarkins) [inline lib/modules/datasource/nuget/v3.ts:309]: Make opt-in, not opt-out

File: lib/modules/datasource/nuget/v3.ts [hunk 8]
```diff
@@ -242,3 +267,65 @@ export async function getReleases(
...
   return dep;
 }
+
+async function getSourceUrlFromNupkgCached(
+  http: Http,
+  packageName: string,
+  packageVersion: string | null,
+  nupkgUrl: string,
+): Promise<string | null> {
+  const cacheKey = `source-url:${nupkgUrl}`;
```
Comment (rarkins) [inline lib/modules/datasource/nuget/v3.ts:277]: Does this URL change every time there's a new release?

I would like it to be cached per _package_ for one week

**Review by rarkins — CHANGES_REQUESTED** [review]

File: lib/modules/datasource/nuget/v3.ts (hunk 8 above)
Comment (viceice) [reply lib/modules/datasource/nuget/v3.ts:277]: yes, that url always includes the version

Comment (viceice) [reply lib/modules/datasource/nuget/v3.ts:277]: we should also do the caching for the nuspec file, which currently also occur on every getReleases call.

Comment (rarkins) [reply lib/modules/datasource/nuget/v3.ts:277]: So the cacheKey should be changed to be based off the registryUrl+packageName, not nupkgUrl

File: docs/usage/self-hosted-experimental.md (hunk 5 above)
Comment (fgreinacher) [reply docs/usage/self-hosted-experimental.md:147]: Also fine for me, maybe I misread the [comment above](https://github.com/renovatebot/renovate/pull/28071#discussion_r1534229452).

File: lib/modules/datasource/nuget/v3.ts (hunk 8 above)
Comment (fgreinacher) [reply lib/modules/datasource/nuget/v3.ts:277]: I am not sure about just taking the package name. It's not too uncommon that a source project changes over time. Depending on which version we look at first we might end up with outdated source URLs being cached for quite some time.

File: lib/modules/datasource/nuget/v3.ts (hunk 6 above)
Comment (fgreinacher) [reply lib/modules/datasource/nuget/v3.ts:238]: Can you explain why you prefer this instead of keeping the data structured?

File: lib/modules/datasource/nuget/v3.ts (hunk 4 above)
Comment (fgreinacher) [reply lib/modules/datasource/nuget/v3.ts:297]: The datasource is mostly forwarding to the v2 or v3 functions at the moment and I don't think it makes sense to add v3-specific logic there.

Should I send a PR to refactor v2 and v3 into classes first?

File: lib/modules/datasource/nuget/v3.ts (hunk 7 above)
Comment (fgreinacher) [reply lib/modules/datasource/nuget/v3.ts]: Done

File: docs/usage/self-hosted-experimental.md (hunk 5 above)
Comment (fgreinacher) [reply docs/usage/self-hosted-experimental.md]: Changed in https://github.com/renovatebot/renovate/pull/28071/commits/a0239dee2b1c178b6ef5b73ac2a569e11a659f09

File: lib/modules/datasource/nuget/v3.ts (hunk 8 above)
Comment (rarkins) [reply lib/modules/datasource/nuget/v3.ts:277]: My experience is that packages in most ecosystems very rarely change source URLs. Most commonly never (i.e. same github repo forever), sometimes 1-2 changes over years, almost never ever "regularly". Is it different for nuget?

File: lib/modules/datasource/nuget/v3.ts (hunk 4 above)
Comment (fgreinacher) [reply lib/modules/datasource/nuget/v3.ts:297]: I have sent https://github.com/renovatebot/renovate/pull/28117

File: lib/modules/datasource/nuget/v3.ts (hunk 8 above)
Comment (fgreinacher) [reply lib/modules/datasource/nuget/v3.ts:277]: I have seen it from time to time, especially with Microsoft/.NET packages. But we can also start with just `registryUrl+packageName` and see how it behaves. Would be easy to adapt if needed.

File: lib/modules/datasource/nuget/v3.ts (hunk 4 above)
Comment (fgreinacher) [reply lib/modules/datasource/nuget/v3.ts]: Switched to cache decorator with ca67215479147747d209ec929afbaa950cb0e6b5

File: lib/modules/datasource/nuget/v3.ts (hunk 8 above)
Comment (fgreinacher) [reply lib/modules/datasource/nuget/v3.ts]: Adapted in ca67215479147747d209ec929afbaa950cb0e6b5

File: lib/modules/datasource/nuget/v3.ts (hunk 6 above)
Comment (fgreinacher) [reply lib/modules/datasource/nuget/v3.ts]: I checked the code base again and the approach is a bit inconsistent. I will adapt this line as suggested, but maybe we want to discuss the desired direction and document it in https://github.com/renovatebot/renovate/tree/main/docs/development?

Comment (rarkins) [reply lib/modules/datasource/nuget/v3.ts]: Logging is inconsistent but my approach is to minimize metadate use in debug messages because it blows out the number of lines in logs. Better to use them only when the data is large, or needed for "machine" parsing (e.g. warnings, errors sent to log systems)

Comment (fgreinacher) [reply lib/modules/datasource/nuget/v3.ts]: @rarkins OK thanks, do you think this is worth to add to the development docs? I can start something, would need a lot of input though.

Comment (rarkins) [reply lib/modules/datasource/nuget/v3.ts]: Definitely. Main thing is that we want _no_ templating in warnings and errors (so they can be aggregated and monitored) but usually templating in debug

Comment (fgreinacher) [reply lib/modules/datasource/nuget/v3.ts]: OK, I have created https://github.com/renovatebot/renovate/pull/28162 to improvement the existing documentation I have just found :)

File: lib/modules/datasource/nuget/index.spec.ts [hunk 9]
```diff
@@ -14,6 +18,9 @@ const hostRules: any = _hostRules;
 
 jest.mock('../../../util/host-rules', () => mockDeep());
 
+jest.mock('../../../util/cache/package');
```
Comment (viceice) [inline lib/modules/datasource/nuget/index.spec.ts:21]:
```suggestion
jest.mock('../../../util/cache/package', () => mockDeep());
```
test perf, will block jest to load all drips recursive to create a mock.

File: lib/modules/datasource/nuget/index.spec.ts [hunk 10]
```diff
@@ -302,6 +309,160 @@ describe('modules/datasource/nuget/index', () => {
       );
     });
 
+    describe('determine source URL from nupkg', () => {
+      beforeEach(() => {
+        GlobalConfig.set({
```
Comment (viceice) [inline lib/modules/datasource/nuget/index.spec.ts:314]: I think we should do a reset in before each for all tests to not cause side effects later

File: lib/modules/datasource/nuget/v3.ts [hunk 11]
```diff
@@ -151,15 +156,23 @@ export class NugetV3Api {
...
+        packageContent,
+      }) => {
         const release: Release = { version: removeBuildMeta(version) };
         if (releaseTimestamp) {
           release.releaseTimestamp = releaseTimestamp;
         }
         if (versioning.isValid(version) && versioning.isStable(version)) {
           latestStable = removeBuildMeta(version);
           homepage = projectUrl ? massageUrl(projectUrl) : homepage;
+          nupkgUrl = packageContent ? massageUrl(packageContent) : null;
```
Comment (viceice) [inline lib/modules/datasource/nuget/v3.ts:175]: we should update message url to accept null and return null if it was passed.

**Review by viceice — COMMENTED** [review]
otherwise LGTM

File: lib/modules/datasource/nuget/v3.ts (hunk 11 above)
Comment (fgreinacher) [reply lib/modules/datasource/nuget/v3.ts]: Adapted in 1a09471d2c382d21e618b5e36ddc195f7e60f131

File: lib/modules/datasource/nuget/index.spec.ts (hunk 10 above)
Comment (fgreinacher) [reply lib/modules/datasource/nuget/index.spec.ts:318]: Done in e7bc6f53c5ac9410d1f5229d560f93211fffe959

File: lib/modules/datasource/nuget/v3.ts [hunk 12]
```diff
@@ -233,4 +258,54 @@ export class NugetV3Api {
...
+      registryUrl: string,
+      packageName: string,
+      _packageVersion: string | null,
+      _nupkgUrl: string,
+    ) => `source-url:${registryUrl}:${packageName}`,
+    ttlMinutes: 10080, // 1 week
+  })
+  async getSourceUrlFromNupkg(
+    http: Http,
+    _registryUrl: string,
```
Comment (viceice) [inline lib/modules/datasource/nuget/v3.ts:275]: remove if unused

**Review by viceice — COMMENTED** [review]
see comment

File: lib/modules/datasource/nuget/v3.ts (hunk 12 above)
Comment (fgreinacher) [reply lib/modules/datasource/nuget/v3.ts:275]: We need it for the cache key :)

**Review by viceice — DISMISSED** [review]

File: lib/modules/datasource/nuget/v3.ts [hunk 13]
```diff
@@ -233,4 +258,54 @@ export class NugetV3Api {
...
+    packageName: string,
+    packageVersion: string | null,
+    nupkgUrl: string,
+  ): Promise<string | null> {
+    // istanbul ignore if: experimental feature
+    if (!process.env.RENOVATE_X_NUGET_DOWNLOAD_NUPKGS) {
+      logger.debug(
+        `Skipping nupkg download because RENOVATE_X_NUGET_DOWNLOAD_NUPKGS is not set.`,
+      );
```
Comment (rarkins) [inline lib/modules/datasource/nuget/v3.ts:284]: Won't this log this debug message once per package for every user not using the option?

Comment (fgreinacher) [reply lib/modules/datasource/nuget/v3.ts:284]: Yes, good point. `logger.once`?

Comment (rarkins) [reply lib/modules/datasource/nuget/v3.ts:284]: In its current state I think it's at risk of alarming people / creating false concern. If you're certain this should be debug then keeping it to `RENOVATE_X_NUGET_DOWNLOAD_NUPKGS is not set` is better so that the majority will just ignore it

**Review by fgreinacher — COMMENTED** [review]
Commented by mistake

File: lib/modules/datasource/nuget/v3.ts (hunk 13 above)
Comment (fgreinacher) [reply lib/modules/datasource/nuget/v3.ts]: Adapted ✅

**Review by viceice — APPROVED** [review]

File: lib/modules/datasource/nuget/v3.ts [hunk 14]
```diff
@@ -233,4 +258,52 @@ export class NugetV3Api {
...
     return dep;
   }
+
+  @cache({
+    namespace: NugetV3Api.cacheNamespace,
+    key: (
+      _http: Http,
+      registryUrl: string,
+      packageName: string,
+      _packageVersion: string | null,
```
Comment (rarkins) [inline lib/modules/datasource/nuget/v3.ts:268]: Doesn't this mean that the cached data will be invalidated each time there's a new release?

Comment (viceice) [reply lib/modules/datasource/nuget/v3.ts:268]: no, it's not used, it's just a required function parameter

**Review by rarkins — APPROVED** [review]
